### ISSUE TITLE
Legacy helm -> istioctl revisions migration path

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#22874]
This issue tracks work and docs needed for getting people off of the legacy helm install onto istioctl using revisions

* Blocker: MCP does not respect revision label
* Blocker: Old pilot holds leader election lock, blocking the root cert disrtibution.
* Blocker: double metrics
* TODO: add docs. As a blog post?

**Comment by howardjohn** [comment]
cc @nrjpoddar

**Comment by nrjpoddar** [comment]
FYI in order to make SDS work in 1.5 without third-party-jwt in the old Helm charts we had to do the following:

1. Add a new Helm option `.global.jwtPolicy` and set it by default to `first-party-jwt`
1. Add ENV variable `JWT_POLICY` in all sidecar and gateway proxies (including the ones added manually in control plane).
1. Guard the `istio-token` volume so that it's only enabled if `jwtPolicy` is `third-party-jwts`
1. Augment the sidecar injector template with changes for items 2 & 3 so that all workloads will be injected with the correct env and volume setting.

**Comment by howardjohn** [comment]
@nrjpoddar unless I am misunderstanding this, that is not related to this work right? This is for migrating from legacy helm charts to new helm charts. That should not require using SDS with the legacy helm charts?

Glad you got it working though

**Comment by nrjpoddar** [comment]
> @nrjpoddar unless I am misunderstanding this, that is not related to this work right? This is for migrating from legacy helm charts to new helm charts. That should not require using SDS with the legacy helm charts?
> 
> Glad you got it working though

Correct! I just didn't have a good place to put this information. Feel free to delete the comment if it's confusing.

**Comment by howardjohn** [comment]
No sounds good, just wanted to verify I wasn't missing something. thanks

**Comment by chadlwilson** [comment]
Hi @howardjohn - am I reading this correctly in concluding that currently there is no migration path for active production workloads (i.e without outage) from a Helm-managed installation to `istioctl` and this ticket is to track work to provide this supported migration path?

Currently we pipeline-automate all of our istio upgrades through multiple environments on different clusters via `helm upgrade --wait` so looking at how we can transition in an automated way.

I've been trying to find documentation of how to migrate or an overview of the current state, but haven't been able to find any, and the existing docs seem to imply that `istioctl upgrade` is only suitable if you installed that way in the first place. It's further complicated in my head by not fully understanding how the installation/upgrade tool is related to `istiod`, mixer-less telemetry and the consolidated control plane, and how best to make all these changes incrementally.

In any case, would the team recommend that users transition from `1.4.7` to `1.5.x` using Helm to prepare us best for any supported migration path that will be instated with `1.6.x`?

**Comment by howardjohn** [comment]
There is not yet a migration path for helm to istioctl, but it will certainly exist in 1.6,which is what this issue is tracking. You can go directly from 1.4 - 1.6 if desired once that is in place. Sorry about some of the confusion, as didn't do a great job around this

**Comment by chadlwilson** [comment]
@howardjohn No need to apologise - you lot are doing some impressive work in a lot of different areas in parallel. Thanks for the clarification!

If there is no reason to hold off (e.g if the Helm `1.5` charts were in a slightly dodgy state due to their deprecation) I think we will try and get onto `1.5` via Helm for the better support for headless services and general philosophy of "change as little as possible as frequently as possible" for migrations.

I imagine migration from `1.5` to `1.6` will be more likely to be heavily tested from a control plane compatibility perspective so this seems a good compromise to reduce risk.

**Comment by howardjohn** [comment]
The upgrade path isn't too different for 1.4 for 1.5 for what it's worth due to https://preliminary.istio.io/blog/2020/multiple-control-planes/ which is how this will be handled. But always encouraged to update to the latest and greatest

**Comment by meganwalker-ibm** [comment]
Hi,

I'm trying to piece some thigns together from this issue and others and trying to make sure we don't end up unsupported. We're in a similar state to Chad above.

Currently we're on a 1.4 helm install. 
As I understand it, we can't currently do a `helm upgrade` to 1.5.x as documented in this issue.

We could hold off until 1.6.x is released, at which point you will support users going from 1.4 to 1.6 without going via 1.5?

That would be a change to your current support statement, atleast for helm upgrades
> Istio does NOT support skip level upgrades. Only upgrades from 1.4 to 1.5 are supported. If you are on an older version, please upgrade to 1.4 first.
[Source](https://istio.io/docs/setup/upgrade/cni-helm-upgrade/)
Can you confirm that that skipping 1.5 will be supported?


We could switch to `helm template` but that won't get us the `istiod` changes, instead leaving us on a Pilot based install. [Source](https://github.com/istio/istio/issues/21648#issuecomment-594458781) Is a pilot based 1.5 well tested/supported?

**Comment by chadlwilson** [comment]
My understanding is that
* `helm upgrade` to `1.5.x` is still supported - the charts still exist and are updated (maybe there are some fixes/flexibility in deployment that is not being resolved in those however?)
* `helm install` for new control plane instances is essentially deprecated
* `helm install` + `istioctl upgrade` is currently unsupported/not working (which is what this ticket is about resolving). :-)

**Comment by meganwalker-ibm** [comment]
Apologies I think I mis spoke on `helm upgrade` to 1.5
AIUI that isn't supported -  https://github.com/istio/istio.io/pull/6725 removed the documentation for `helm upgrade` in part because https://github.com/istio/istio/issues/21648 showed that `helm upgrade` to 1.5 was broken

**Comment by howardjohn** [comment]
Our support policy is sort of irrelevant here, to some extent. We are adding support for multiple independent control planes, so you could move from Istio 1.1 to Istio 1.6 (in theory - don't actually try that though).

The recommended path would be to upgrade to 1.5 first then do the canary deployment to 1.6. However since apparently `helm upgrade` is broken by helms restrictions that may not work with tiller (although I think you can force it). If you use template you certainly can.

**Comment by chadlwilson** [comment]
It seems you are right on support for `helm upgrade` @meganwalker-ibm ; thanks for the correction.

Frankly I'm a bit shocked by the removal of `helm upgrade` - it seems a bit anti-user; because unless I missed something it doesn't look like the decision was made based on user stats/usage patterns. Using `helm template | kubectl apply` isn't the same thing; when automating you have to construct your own `kubectl` commands coupled with the internal structure of the thing you're deployment (wait things to wait on, which rollouts to check status of, which jobs need to run...) to verify the state of every active deployed resource/component to validate; rather than what Helm gives you with `--wait --cleanup-on-fail` etc.

I see what I consider to be a bit of FUD in the PR about tiller security (irrelevant since tiller is gone in Helm 3, and client-side tiller is totally possible in Helm 2 via the helm-tiller plugin) but this looks a bit rash to me personally.

Later I will see if `helm upgrade` is indeed broken for us though.

**Comment by howardjohn** [comment]
Basic test of 1.5 legacy helm -> master is working now.

I tested 1.2 -> master as well. It sort of works, but the problem is the validating webhook. The EnvoyFilters for telemetry v2 are now invalid according to the 1.2 validation logic. To get around this you would need to disable telemetry v2 or disable validation on galley (`--enable-validation=false` and delete the webhook).

**Comment by nrjpoddar** [comment]
Verified the new "canary upgrade" from an Helm 3 installed version of 1.4.6 -> Helm 3 installed version of 1.6.alpha-46477282051c75b9005f64fe89e34d4bb99699ae. These were the issues found:

1. `helm template --namespace istio-system istio-base ./manifests/charts/base/ --set revision=canary` has many resources overlapping w/istio 1.4.6 install. Helm 3 will not install them and give out the following error:
  `Error: rendered manifests contain a resource that already exists. Unable to continue with install: 
    existing resource conflict: namespace: istio-system, name: istio-pilot-service-account, 
    existing_kind: /v1, Kind=ServiceAccount, new_kind: /v1, Kind=ServiceAccount`

1. Another issue is that the 1.4.6 control plane rejects the telemetry v2 EnvoyFilter installed by 1.6 canary:
  `admission webhook "pilot.validation.istio.io" denied the request: configuration is invalid: Envoy 
    filter: unknown message type "udpa.type.v1.TypedStruct"`
1. Deleting canary version ends up removing a lot of currently installed resources which makes the cluster unusable.

Possible solutions:

1. Move the ClusterRole/ClusterRoleBinding/ServiceAccount from "base" chart to components that need them e.g. "istio-control/istio-discovery" and add suffix based on "revision"
1. For old supported Istio releases (1.5, 1.4) add an option to disable EnvoyFilter webhook validation as old control planes don't honor rev labels and don't understand new EnvoyFilter schema/types.
1. Install multiple webhooks per revisions i.e. chain them so that new revisions validate the resources they know about along with old control plane versions
1. The "base" chart need to not install the old CRDs if revision is set. This is to allow Helm3 to install base chart for canary revisions.

**Comment by nrjpoddar** [comment]
CC @sdake for visibility.

**Comment by craigbox** [comment]
I can help with the blog post.

**Comment by jaygorrell** [comment]
Reading through this, it sounds like 1.4 -> 1.5 using `helm template` (not `helm upgrade`) should work, and then 1.5 -> 1.6 as a migration to `istioctl` would be next. However, I tried that first step today and had all kinds of connectivity issues (and of course no `istiod`).

I can wait on the blog post but wanted to call this out as something that should be made _very_ clear for users when it's written.

**Comment by nrjpoddar** [comment]
> Reading through this, it sounds like 1.4 -> 1.5 using `helm template` (not `helm upgrade`) should work, and then 1.5 -> 1.6 as a migration to `istioctl` would be next. However, I tried that first step today and had all kinds of connectivity issues (and of course no `istiod`).
> 
> I can wait on the blog post but wanted to call this out as something that should be made _very_ clear for users when it's written.

@jaygorrell Helm support was dropped in 1.5 if you want to use the new charts (install/kubernetes/operator which gives you istiod). However, Helm works in 1.5 with the old chart (install/kubernetes/helm/) which doesn't give you istiod but the old microservices control plane. There are other subtle differences between the old and new charts so based on your use case one or the other might be better.

**Comment by jaygorrell** [comment]
Thanks @nrjpoddar, I missed the new location for those. The old charts (install/kubernetes/helm) gave me a broken cluster, though -- sidecars stopped receiving config from pilot it seemed, and the ingress-gateway deployment failed to come up. I didn't spend much more time debugging before rolling back to 1.4 though.

But you'd say those really should work? I've been upgrading since 1.1 and this was the first issue like that. I'll try again if it's well tested and should work.

**Comment by howardjohn** [comment]
Regarding @nrjpoddar's comments:

It seems there are two usages for revisions:
1. Deploying two revisions of the same version.
2. Using a revision to upgrade between versions.

I expect that how we configure our install will be different based on which of those we plan for. (2) is likely much more common, so we should optimize for this case.

The base chart was meant as a way to separate cluster resources (installed by cluster admin) from others. This was mostly accomplished, except the mutating webhook. The intent was this was 1 per cluster, and not tied to a revision.

This has some issues, as Neeraj mentioned:
* If a role changes, we have some issues. 1.x and 1.y may have incompatible requirements on what is required in the roles
* Singleton validation webhook has issues. If I apply a "global" config (no `istio.io/rev` label), I would expect it to work for all existing revisions. If I apply a rev specific config, I only care that it works for that revision (and should NOT validate it against other revisions). This is a bit tricky because prior to 1.5 istio.io/rev label is not respected
* Reusing the config names, breaking `helm install`
* A bit hard to handle deletion. Not a huge blocker though as you can just delete the `pilot` chart, not others.

If we move ClusterRole/ClusterRoleBinding/ServiceAccount/Validation to the pilot component and suffix with revision, we mostly solve case (2). However, case (1) is made worse, as we don't have the privilege split, and we do multiple validation calls for no reason.

**Comment by nrjpoddar** [comment]
> Regarding @nrjpoddar's comments:
> 
> It seems there are two usages for revisions:
> 
> 1. Deploying two revisions of the same version.
> 2. Using a revision to upgrade between versions.
> 
> I expect that how we configure our install will be different based on which of those we plan for. (2) is likely much more common, so we should optimize for this case.
> 
Agree about ensuring we make use case 2 work as it's the most asked feature. Making 1 work with requiring some more options/commands from users is not bad IMO.

> The base chart was meant as a way to separate cluster resources (installed by cluster admin) from others. This was mostly accomplished, except the mutating webhook. The intent was this was 1 per cluster, and not tied to a revision.
> 
> This has some issues, as Neeraj mentioned:
> 
> * If a role changes, we have some issues. 1.x and 1.y may have incompatible requirements on what is required in the roles
> * Singleton validation webhook has issues. If I apply a "global" config (no `istio.io/rev` label), I would expect it to work for all existing revisions. If I apply a rev specific config, I only care that it works for that revision (and should NOT validate it against other revisions). This is a bit tricky because prior to 1.5 istio.io/rev label is not respected
> * Reusing the config names, breaking `helm install`
> * A bit hard to handle deletion. Not a huge blocker though as you can just delete the `pilot` chart, not others.
> 
> If we move ClusterRole/ClusterRoleBinding/ServiceAccount/Validation to the pilot component and suffix with revision, we mostly solve case (2). However, case (1) is made worse, as we don't have the privilege split, and we do multiple validation calls for no reason.

Architecturally we shouldn't separate use case 1 & 2 as that indicates we haven't designed it well. Here are some fundamental ideas that should apply here:

* If two revisions should have separate ownership for resources let's ensure we have that via labels/name suffixing/etc.
* If we share "true" singletons (like CRDs) these are the options:
  + If the shared resources by two revisions are identical + new revision has new additions, then you can simply guard the shared resource install and declaratively make sure the new version only own additions.
    + If declarative guards become too complex, you can shield the shared resources by an imperative logic (like a job) which in turn will do the right thing based on cluster state.
  + If the shared resources by two revisions themselves change (e.g. OpenAPI specification for a CRD has changed), you need a more complex ownership pattern:
    + You probably need jobs which create/augment these resources and add label with a list of revisions that know about them. You can old delete/gc when the "list" of owners get empty.

**Comment by howardjohn** [comment]
@linsun also brought up another use case where they want just case+mutation webhook since they will have the Istiod in another cluster

**Comment by linsun** [comment]
subscribe - interested in this because I'm working on https://github.com/istio/istio/issues/23028 where i want to expose caBundle as a config option for remote cluster where istiod isn't enabled but i want my mutating webhook configuration to be there.

**Comment by howardjohn** [comment]
Also Helm3 has first class support for `crds` folder which may help some of the resource conflict errors (it will fail if the crd is in template/ but not crd/)

**Comment by linsun** [comment]
Another resource I need on remote cluster is istio-ca-root-cert configmap, which is currently configured in istio-discovery.

**Comment by howardjohn** [comment]
istio-ca-root-cert  is created at runtime so it is a bit orthogonal to the
install structure I think

**Comment by nrjpoddar** [comment]
how do I mark this as a blocker for 1.6?

**Comment by howardjohn** [comment]
its already marked as a blocker through zenhub.

**Comment by shwsingh** [comment]
@howardjohn  Can you please provide ETA ?

**Comment by shwsingh** [comment]
@howardjohn  Can you please provide ETA since it's a release blocker ?

**Comment by howardjohn** [comment]
We have a PR: https://github.com/istio/istio/pull/23194 so blocked by review of this PR. its only 5 lines so Ill optimistically say today

**Comment by brian-avery** [comment]
@howardjohn can this be closed? 23194 looks like it's merged?

**Comment by howardjohn** [comment]
I think the last remaining step is a quick guide in te upgrade notes

**Comment by cbron** [comment]
This approach is apparently not supported on operator installs in 1.6: https://github.com/istio/istio/issues/23479

**Comment by howardjohn** [comment]
Upgrade notes in https://github.com/istio/istio.io/pull/7275/files
In cluster operator in https://github.com/istio/istio/issues/23479

So I think this is safe to close out for now. If anyone does experience issues during upgrade please open an issue and mention me. Thanks!

**Comment by tmshort** [comment]
@howardjohn 
> The upgrade path isn't too different for 1.4 for 1.5 for what it's worth due to https://preliminary.istio.io/blog/2020/multiple-control-planes/ which is how this will be handled. But always encouraged to update to the latest and greatest

I'm looking at this page, and there's an iframe'd Google Sheets document that is asking for permission to access, and it's failing. I don't know if it's a Google Drive problem or what. Regardless, every time I want to access a "public" istio Google document, I need to ask for access. Couldn't these documents just be made publicly viewable for everyone without people asking for access?

**Comment by howardjohn** [comment]
@tmshort see https://github.com/istio/istio/issues/20282.

**Comment by tmshort** [comment]
Thanks, but it doesn't help that I asking for access is failing for me now. I will comment there.
